include/EASTL/sort.h

```diff
@@ -202,80 +202,189 @@ namespace eastl
+	                EASTL_VALIDATE_COMPARE(!(*iCurrent < temp)); // Validate that the compare function is sane.
+	                *iNext = *iCurrent;
+	            }
+	
+	            *iNext = temp;
+	        }
+	    }
 	} // insertion_sort
-
+	
+	
+	
+	///modified insertion sort that fixes the decrement problem
+	///
+	///also it seems that:
+	///for (...; ...; iRight=iLeft--)
+	///can be slighlty faster than:
+	///for (...; ...; --iLeft, --iRight)
+	///
+	///inner loop changed(after above change)
+	///from this:	for(--iLeft; iRight!=first && temp<*iLeft  ; iRight=iLeft--){}
+	///  to this:	for(       ; iRight!=first && temp<*--iLeft; iRight=iLeft  ){}
+	
+	//compare functor version
+	template<typename BidirectionalIterator, typename Compare>//Bidirectional Iterator, though something like a list shouldn't use this
+	void insertion_sort_new(BidirectionalIterator const first, BidirectionalIterator const last, Compare cmp)
+	{
+	    typedef typename eastl::iterator_traits<BidirectionalIterator>::value_type value_type;
+	
+	    if (first!=last)//if range not empty
+	    {
+	        BidirectionalIterator iLeft, iRight, iSorted=first;
+	        for (++iSorted; iSorted!=last; ++iSorted)//an array of length 1 is already sorted
+	        {
+	            const value_type temp(*iSorted);
+	
+	            for (iLeft=iRight=iSorted; iRight!=first && cmp(temp, *--iLeft); iRight=iLeft)
+	            {
+	                EASTL_VALIDATE_COMPARE(!cmp(*iLeft, temp));//validate cmp is sane by checking other way
+	                *iRight=*iLeft;
+	            }
+	
+	            *iRight=temp;
+	        }//outer loop
+	    }
+	}//insertion_sort (Compare)
+	
+	template<typename BidirectionalIterator>//default less-than version
+	void insertion_sort_new(BidirectionalIterator const first, BidirectionalIterator const last)
+	{
+	    typedef typename eastl::iterator_traits<BidirectionalIterator>::value_type value_type;
+	
+	    if (first!=last)//if range not empty
+	    {
+	        BidirectionalIterator iLeft, iRight, iSorted=first;
+	        for (++iSorted; iSorted!=last; ++iSorted)//an array of length 1 is already sorted
+	        {
+	            const value_type temp(*iSorted);
+	
+	            for (iLeft=iRight=iSorted; iRight!=first && temp<*--iLeft; iRight=iLeft)
+	            {
+	                EASTL_VALIDATE_COMPARE(!(*iLeft<temp));//validate cmp is sane by checking other way
+	                *iRight=*iLeft;
+	            }
+	
+	            *iRight=temp;
+	        }//outer loop
+	    }
+	}//insertion_sort (less than)
+	
+	///This following Merge Sort or something similar is worth considering compared to the current version.
+	///First, the buffer only needs a length of half the data, not the full length as it does now.
+	///Second, it starts by using Insertion Sort on small arrays, which noticeably improves the performance. The full sort is still stable.
+	
+	//the following is like a normal merge, but nothing remains to be copied if the left range is exhausted first
```

Comment:
I'd suggest taking the merge sort changes and submitting them in another PR so they can discussed separately from the insertation sort changes.